### ISSUE TITLE
Your programming language sucks

### ISSUE THREAD

**Issue by capr** [capr/blog#20]
## Garbage collection

Your language sucks because it either has a gc or it doesn't have a gc. This is why the most complex programs in existence today (games, CAD software, etc.) need to use two languages: one for the engine and one for the application logic and UI.

## Exceptions

Your language sucks because it doesn't have a way to propagate errors down the call stack and catch them later and unwind the stack. 

Because your language doesn't have exceptions, all my networking code written in your language sucks. Now I cannot write a web server that closes the connection automatically and frees the memory associated with it when I/O fails, and handle that in a single place. Instead, I have to propagate the error condition myself manually through the entire network, protocol and application code.

Exceptions are almost never needed in a program, I'm down with that. But just because Java created a culture of exceptions everywhere, just because C++ botched it with RAII, doesn't mean that you have to throw the baby with the bathwater. In networking code, the pattern of recoverable failure which is what exceptions are addressing, _is_ the optimal coding pattern. You'll have resource leaks, you say? Figure it out so you don't. It doesn't have to be RAII and it doesn't have to involve the heap.

## RTTI

TODO

## Virtual properties

Because your language doesn't have a way to virtualize field access, all my UI code written in your langauge sucks because I have to write x = a.get_this() and b.set_that(y) instead of x = a.this and b.that = y.

## Operator overloading

Because your language doesn't have operator overloading, all my math code written in your language that involves 2D or 3D points, matrices, quaternions, etc. sucks because I have to write mat_add(a, mat_mul(b, c)) instead of a + b * c.

## Closures

TODO

## Coroutines

Like exceptions, coroutines are almost never needed in a program.
TODO

## Variable shadowing

This is a minor point.
TODO

## Conclusion

In conclusion, your programming language sucks because it's opinionated. Because you think you have a duty to protect the users from themselves. Because you think you have to penalize bad behavior. Because you don't see the value in a language feature because you haven't used it yourself.  Because you're a systems programmer who doesn't understand the value of dynamic languages. Because you're a game programmer who thinks complexity is on a single axis so there can only be more or less of it (and of course games have the most).

**Comment by 4zv4l** [comment]
What about Nim ? It seems to check most of what you said. Still sucks ? xD

**Comment by capr** [comment]
Yep, Nim does check a lot of boxes. Not a criticism of Nim, but my programs tend to want to be 80% dynamically typed, 20% statically typed. If I have to write that 80% in a statically typed lang, even one as terse as Nim regarding type declarations, it would still slow me down a bunch and feel like a straitjacket (kinda like golang). I want my hashmaps-with-everything to just work, it's where I keep most of the data (why don't they btw? since Nim has gc and generics, that should be enough). I only need typed buffers and control over allocations for 20% of the program.
